services/messages-service.ts

```diff
@@ -0,0 +1,88 @@
+///<reference path="../.d.ts"/>
+"use strict";
+
+import * as util from "util";
+import * as path from "path";
+
+export class MessagesService implements IMessagesService {
+	private _pathsToMessageJsonFiles: string[] = null;
+	private _messageJsonFilesContentsCache: any[] = null;
+
+	private get pathToDefaultMessageJson(): string {
+		return path.join(path.dirname(__dirname), "resources", "messages", "errorMessages.json");
+	}
+
+	private get messageJsonFilesContents(): any[] {
+		if (!this._messageJsonFilesContentsCache || !this._messageJsonFilesContentsCache.length) {
+			this.refreshMessageJsonContentsCache();
+		}
+
+		return this._messageJsonFilesContentsCache;
+	}
+
+	constructor(private $fs: IFileSystem) {
+		this._pathsToMessageJsonFiles = [this.pathToDefaultMessageJson];
+	}
+
+	public get pathsToMessageJsonFiles(): string[] {
+		if (!this._pathsToMessageJsonFiles) {
+			throw new Error("No paths to message json files provided.");
+		}
+
+		return this._pathsToMessageJsonFiles;
+	}
+
+	public set pathsToMessageJsonFiles(pathsToMessageJsonFiles: string[]) {
+		this._pathsToMessageJsonFiles = pathsToMessageJsonFiles.concat(this.pathToDefaultMessageJson);
+		this.refreshMessageJsonContentsCache();
+	}
+
+	public getMessage(id: string, ...args: string[]): string {
+		let keys = id.split("."),
+			result = this.getFormatedMessage(id, ...args);
+
+		_.each(this.messageJsonFilesContents, jsonFileContents => {
+			let messageValue = this.getMessageFromJsonRecursive(keys, jsonFileContents, 0);
+			if (messageValue) {
+				result = this.getFormatedMessage(messageValue, ...args);
+				return false;
+			}
+		});
+
+		return result;
+	}
+
+	private getMessageFromJsonRecursive(keys: string[], jsonContents: any, index: number): string {
+		if (index >= keys.length) {
+			return null;
+		}
+
+		let jsonValue = jsonContents[keys[index]];
+		if (!jsonValue) {
+			return null;
+		}
+
+		if (typeof jsonValue === "string") {
+			return jsonValue;
+		}
+
+		return this.getMessageFromJsonRecursive(keys, jsonValue, index + 1);
+	}
+
+	private refreshMessageJsonContentsCache(): void {
+		this._messageJsonFilesContentsCache = [];
+		_.each(this.pathsToMessageJsonFiles, path => {
+			if (!this.$fs.exists(path).wait()) {
+				throw new Error("Message json file " + path + " does not exist.");
```

Comment:
@teobugslayer I've chosen `new Error()` out of fear of maximum call stack. Our `$errors` class resolves this `messageService` when getting the message which should be printed on the screen. Hypothetically we could end up in a bottomless recursion